File: apps/mobile/lib/account/device_wrap.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:crypto/crypto.dart' as crypto;
import 'package:pointycastle/export.dart';
import 'package:x25519/x25519.dart' as x25519;

/// Wire format must match `@auvora/vault-crypto` (`auvora-device-wrap-v1`).
const String kDeviceWrapAlg = 'auvora-device-wrap-v1';

class DeviceKeyPair {
  const DeviceKeyPair({required this.publicKey, required this.privateKey});

  final String publicKey;
  final String privateKey;
}

class DeviceWrappedVaultKey {
  const DeviceWrappedVaultKey({
    required this.algorithmId,
    required this.ciphertext,
    required this.nonce,
    required this.ephemeralPublicKey,
    required this.aad,
  });

  final String algorithmId;
  final String ciphertext;
  final String nonce;
  final String ephemeralPublicKey;
  final String aad;

  Map<String, Object?> toJson() => {
        'algorithmId': algorithmId,
        'ciphertext': ciphertext,
        'nonce': nonce,
        'ephemeralPublicKey': ephemeralPublicKey,
        'aad': aad,
      };

  factory DeviceWrappedVaultKey.fromJson(Map<String, dynamic> json) => DeviceWrappedVaultKey(
        algorithmId: (json['algorithmId'] ?? kDeviceWrapAlg).toString(),
        ciphertext: (json['ciphertext'] ?? '').toString(),
        nonce: (json['nonce'] ?? '').toString(),
        ephemeralPublicKey: (json['ephemeralPublicKey'] ?? '').toString(),
        aad: (json['aad'] ?? '').toString(),
      );
}

String _toBase64(Uint8List bytes) => base64Encode(bytes);

Uint8List _fromBase64(String value) => Uint8List.fromList(base64Decode(value));

Uint8List _randomBytes(int length) {
  final rnd = Random.secure();
  return Uint8List.fromList(List<int>.generate(length, (_) => rnd.nextInt(256)));
}

/// RFC5869 HKDF-SHA256 (extract+expand), matching Node device-wrap.
Uint8List hkdfSha256({
  required Uint8List ikm,
  required Uint8List salt,
  required String info,
  required int length,
}) {
  final prk = crypto.Hmac(crypto.sha256, salt).convert(ikm).bytes;
  final infoBytes = utf8.encode(info);
  final blocks = <Uint8List>[];
  var prev = Uint8List(0);
  var counter = 1;
  while (blocks.fold<int>(0, (sum, b) => sum + b.length) < length) {
    final input = Uint8List.fromList(<int>[...prev, ...infoBytes, counter]);
    final block = Uint8List.fromList(
      crypto.Hmac(crypto.sha256, prk).convert(input).bytes,
    );
    blocks.add(block);
    prev = block;
    counter += 1;
  }
  final out = Uint8List.fromList(blocks.expand((b) => b).toList());
  return Uint8List.sublistView(out, 0, length);
}

Uint8List _x25519SharedSecret(Uint8List privateKey, Uint8List publicKey) {
  return Uint8List.fromList(x25519.X25519(privateKey, publicKey));
}

/// AES-256-GCM for device-wrap: `ciphertext || tag(16)`, nonce supplied separately.
Uint8List _deviceWrapAesGcmEncrypt({
  required Uint8List key,
  required Uint8List plaintext,
  required String aad,
  required Uint8List nonce,
}) {
  final cipher = GCMBlockCipher(AESEngine())
    ..init(
      true,
      AEADParameters(KeyParameter(key), 128, nonce, utf8.encode(aad)),
    );
  final processed = cipher.process(plaintext);
  if (processed.length < 16) {
    throw StateError('AES-GCM encrypt produced invalid output');
  }
  final encrypted = processed.sublist(0, processed.length - 16);
  final tag = processed.sublist(processed.length - 16);
  return Uint8List.fromList(<int>[...encrypted, ...tag]);
}

Uint8List _deviceWrapAesGcmDecrypt({
  required Uint8List key,
  required Uint8List payload,
  required String aad,
  required Uint8List nonce,
}) {
  if (payload.length < 17) {
    throw StateError('Invalid device wrap ciphertext');
  }
  final encrypted = payload.sublist(0, payload.length - 16);
  final tag = payload.sublist(payload.length - 16);
  final cipher = GCMBlockCipher(AESEngine())
    ..init(
      false,
      AEADParameters(KeyParameter(key), 128, nonce, utf8.encode(aad)),
    );
  return cipher.process(Uint8List.fromList(<int>[...encrypted, ...tag]));
}

Uint8List _deviceWrapSalt(String requestId) {
  final input = utf8.encode('$kDeviceWrapAlg|$requestId');
  return Uint8List.fromList(crypto.sha256.convert(input).bytes);
}

String _deviceWrapAad({required String ownerUserId, required String requestId}) =>
    '$kDeviceWrapAlg|$ownerUserId|$requestId';

/// Generate an X25519 keypair for a requesting device (client-side only).
DeviceKeyPair generateDeviceRecoveryKeyPair() {
  final pair = x25519.generateKeyPair();
  return DeviceKeyPair(
    publicKey: _toBase64(Uint8List.fromList(pair.publicKey)),
    privateKey: _toBase64(Uint8List.fromList(pair.privateKey)),
  );
}

/// Wrap a 32-byte vault key for a requesting device public key (trusted device).
DeviceWrappedVaultKey wrapVaultKeyForDevice({
  required Uint8List vaultKey,
  required String recipientPublicKey,
  required String requestId,
  required String ownerUserId,
}) {
  if (vaultKey.length != 32) {
    throw ArgumentError('Vault key must be 32 bytes');
  }
  final recipientPk = _fromBase64(recipientPublicKey);
  final ephemeral = x25519.generateKeyPair();
  final shared = _x25519SharedSecret(
    Uint8List.fromList(ephemeral.privateKey),
    recipientPk,
  );
  final salt = _deviceWrapSalt(requestId);
  final aad = _deviceWrapAad(ownerUserId: ownerUserId, requestId: requestId);
  final aesKey = hkdfSha256(ikm: shared, salt: salt, info: 'auvora-device-wrap', length: 32);
  final nonce = _randomBytes(12);
  final ciphertext = _deviceWrapAesGcmEncrypt(
    key: aesKey,
    plaintext: vaultKey,
    aad: aad,
    nonce: nonce,
  );
  return DeviceWrappedVaultKey(
    algorithmId: kDeviceWrapAlg,
    ciphertext: _toBase64(ciphertext),
    nonce: _toBase64(nonce),
    ephemeralPublicKey: _toBase64(Uint8List.fromList(ephemeral.publicKey)),
    aad: aad,
  );
}

/// Unwrap a vault key on the requesting device using its private key.
Uint8List unwrapVaultKeyForDevice({
  required DeviceWrappedVaultKey wrapped,
  required String recipientPrivateKey,
  required String requestId,
  required String ownerUserId,
}) {
  if (wrapped.algorithmId != kDeviceWrapAlg) {
    throw StateError('Unsupported device wrap algorithm');
  }
  final expectedAad = _deviceWrapAad(ownerUserId: ownerUserId, requestId: requestId);
  if (wrapped.aad != expectedAad) {
    throw StateError('Device wrap AAD mismatch');
  }
  final recipientSk = _fromBase64(recipientPrivateKey);
  final ephemeralPk = _fromBase64(wrapped.ephemeralPublicKey);
  final shared = _x25519SharedSecret(recipientSk, ephemeralPk);
  final salt = _deviceWrapSalt(requestId);
  final aesKey = hkdfSha256(ikm: shared, salt: salt, info: 'auvora-device-wrap', length: 32);
  final nonce = _fromBase64(wrapped.nonce);
  final payload = _fromBase64(wrapped.ciphertext);
  return _deviceWrapAesGcmDecrypt(key: aesKey, payload: payload, aad: wrapped.aad, nonce: nonce);
}

File: packages/vault-crypto/src/device-wrap.ts
/**
 * Device-to-device vault-key wrapping (X25519 ECDH + HKDF-SHA256 + AES-256-GCM).
 * Server relays ciphertext only — never sees the vault key plaintext.
 */
import { createCipheriv, createDecipheriv, createHash, createHmac, randomBytes } from 'node:crypto';
import { x25519 } from '@noble/curves/ed25519.js';

export const DEVICE_WRAP_ALG = 'auvora-device-wrap-v1';

export type DeviceKeyPair = {
  publicKey: string;
  privateKey: string;
};

export type DeviceWrappedVaultKey = {
  algorithmId: typeof DEVICE_WRAP_ALG;
  ciphertext: string;
  nonce: string;
  ephemeralPublicKey: string;
  aad: string;
};

function toB64(buf: Uint8Array): string {
  return Buffer.from(buf).toString('base64');
}

function fromB64(value: string): Buffer {
  return Buffer.from(value, 'base64');
}

function hkdfSha256(ikm: Uint8Array, salt: Uint8Array, info: string, length: number): Buffer {
  const prk = createHmac('sha256', Buffer.from(salt)).update(Buffer.from(ikm)).digest();
  const infoBuf = Buffer.from(info, 'utf8');
  const blocks: Buffer[] = [];
  let prev = Buffer.alloc(0);
  let counter = 1;
  while (Buffer.concat(blocks).length < length) {
    const block = createHmac('sha256', prk)
      .update(Buffer.concat([prev, infoBuf, Buffer.from([counter])]))
      .digest();
    blocks.push(block);
    prev = block;
    counter += 1;
  }
  return Buffer.concat(blocks).subarray(0, length);
}

/** Generate an X25519 keypair for a requesting device (client-side only). */
export function generateDeviceRecoveryKeyPair(): DeviceKeyPair {
  const privateKey = x25519.utils.randomPrivateKey();
  const publicKey = x25519.getPublicKey(privateKey);
  return { publicKey: toB64(publicKey), privateKey: toB64(privateKey) };
}

/**
 * Wrap a 32-byte vault key for a requesting device public key.
 * Approving (trusted) device only — never run on server with plaintext vault keys.
 */
export function wrapVaultKeyForDevice(args: {
  vaultKey: Uint8Array;
  recipientPublicKey: string;
  requestId: string;
  ownerUserId: string;
}): DeviceWrappedVaultKey {
  if (args.vaultKey.length !== 32) {
    throw new Error('Vault key must be 32 bytes');
  }
  const recipientPk = fromB64(args.recipientPublicKey);
  const ephemeralSk = x25519.utils.randomPrivateKey();
  const ephemeralPk = x25519.getPublicKey(ephemeralSk);
  const shared = x25519.getSharedSecret(ephemeralSk, recipientPk);
  const salt = createHash('sha256').update(`${DEVICE_WRAP_ALG}|${args.requestId}`).digest();
  const aad = `${DEVICE_WRAP_ALG}|${args.ownerUserId}|${args.requestId}`;
  const aesKey = hkdfSha256(shared, salt, 'auvora-device-wrap', 32);
  const nonce = randomBytes(12);
  const cipher = createCipheriv('aes-256-gcm', aesKey, nonce);
  cipher.setAAD(Buffer.from(aad, 'utf8'));
  const encrypted = Buffer.concat([cipher.update(Buffer.from(args.vaultKey)), cipher.final()]);
  const tag = cipher.getAuthTag();
  return {
    algorithmId: DEVICE_WRAP_ALG,
    ciphertext: toB64(Buffer.concat([encrypted, tag])),
    nonce: toB64(nonce),
    ephemeralPublicKey: toB64(ephemeralPk),
    aad,
  };
}

/** Unwrap a vault key on the requesting device using its private key. */
export function unwrapVaultKeyForDevice(args: {
  wrapped: DeviceWrappedVaultKey;
  recipientPrivateKey: string;
  requestId: string;
  ownerUserId: string;
}): Buffer {
  if (args.wrapped.algorithmId !== DEVICE_WRAP_ALG) {
    throw new Error('Unsupported device wrap algorithm');
  }
  const expectedAad = `${DEVICE_WRAP_ALG}|${args.ownerUserId}|${args.requestId}`;
  if (args.wrapped.aad !== expectedAad) {
    throw new Error('Device wrap AAD mismatch');
  }
  const recipientSk = fromB64(args.recipientPrivateKey);
  const ephemeralPk = fromB64(args.wrapped.ephemeralPublicKey);
  const shared = x25519.getSharedSecret(recipientSk, ephemeralPk);
  const salt = createHash('sha256').update(`${DEVICE_WRAP_ALG}|${args.requestId}`).digest();
  const aesKey = hkdfSha256(shared, salt, 'auvora-device-wrap', 32);
  const nonce = fromB64(args.wrapped.nonce);
  const payload = fromB64(args.wrapped.ciphertext);
  if (payload.length < 17) {
    throw new Error('Invalid device wrap ciphertext');
  }
  const encrypted = payload.subarray(0, payload.length - 16);
  const tag = payload.subarray(payload.length - 16);
  const decipher = createDecipheriv('aes-256-gcm', aesKey, nonce);
  decipher.setAAD(Buffer.from(args.wrapped.aad, 'utf8'));
  decipher.setAuthTag(tag);
  return Buffer.concat([decipher.update(encrypted), decipher.final()]);
}
